File: graph.cpp
#include "graph.h"

using namespace std;

Graph::Graph(vector<vector<pair<int,double>>> &a) {
  n_nodes = a.size();
  adj = a;
}

Graph::Graph(vector<Edge> edges) {
  n_nodes = 0;
  for (Edge e : edges) {
    n_nodes = max(n_nodes, e.node1 + 1);
    n_nodes = max(n_nodes, e.node2 + 1);

    while (adj.size() < n_nodes) adj.push_back(vector<pair<int,double>>());
    adj[e.node1].push_back(make_pair(e.node2,e.weight));
    adj[e.node2].push_back(make_pair(e.node1, e.weight));
  }
}

void Graph::print() {
  for (int i = 0; i < adj.size(); i++) {
      cout << i << ": ";
      for (pair<int, double> conn : adj[i]) {
        cout << "(" << conn.first << ", " << conn.second << ") ";
      }
      cout << endl;
  }
}

void Graph::remove_edge(int node, int index) {
  if (node >= adj.size() || index >= adj[node].size()) return;
  int other_node = adj[node][index].first;
  pair<int, double> to_find = make_pair(node, adj[node][index].second);
  adj[other_node].erase(find(adj[other_node].begin(), adj[other_node].end(), to_find));
  adj[node].erase(adj[node].begin() + index);
}

File: Mini Scratch/src/Graph.java
import java.util.*;
public class Graph {
	int number; // number of nodes
	ArrayList<Pair>[] adj;
	
	public Graph (ArrayList<Pair>[] adj) {
		number = adj.length;
		this.adj = adj;
	}
	
	public Graph (ArrayList<Edge> edges) {
		number = 0;
		for (Edge e: edges) {
			number = Math.max(number, e.n1+1);
			number = Math.max(number, e.n2+1);
		}
		adj = new ArrayList[number];
		for(int i=0; i < number; i++) {
			adj[i] = new ArrayList<Pair>();
		}
		for(Edge e:edges) {
			adj[e.n1].add(new Pair(e.n2,e.w));
			adj[e.n2].add(new Pair(e.n1,e.w));
		}
	}

	public void removeEdge(int node, int index) {
		if(node >= adj.length || index >=adj[node].size())
			return;
		int v1 = node;
		int v2 = adj[node].get(index).a;
		double w2 = adj[node].get(index).b;
		
		// remove the edge twice in both direction
		adj[node].remove(index);
		adj[v2].remove(adj[v2].indexOf(new Pair(v1,w2)));
		
	}

	public void dump() {
		for(int i=0; i<adj.length; i++) {
			System.out.println( "node " +i + " adjacent to:");
			for(Pair p: adj[i]) {
				System.out.println(p.a +" weight: "+ p.b);
			}
		}
	}
	
//	@Override
//	public String toString() {
////		return "Graph [number=" + number + ", adj=" + Arrays.toString(adj) + "]";
//		String s ="";
//		for (ArrayList<Pair> p: adj) {
//			s+
//		}
//		return s;
//	}
}
